Source/qol/itemlabels.cpp

```diff
@@ -0,0 +1,170 @@
+/**
+* @file itemlabels.cpp
+*
+* Adds Item Labels QoL feature
+*/
+
+#include "common.h"
+#include "control.h"
+#include "itemlabels.h"
+#include "inv.h"
+#include "render.h"
+#include "gmenu.h"
+#include "cursor.h"
+
+#include <unordered_set>
+#include <vector>
+#include <string>
+
+namespace devilution {
+
+namespace {
+
+struct itemLabelQueue {
+	int itemID, row, col, x, y, width, height;
+	text_color color;
+	std::string text;
+};
+
+std::vector<itemLabelQueue> labelQ;
+int drawMinX;
+int drawMaxX;
+bool altPressed = false;
+int generatingLabelsState = 0; // 0 - waiting for the game to start, 1 = started generating, 2 = finished generating
+bool isLabelHighlighted = false;
+int labelCenterOffsets[ITEMTYPES];
+bool invertHighlightToggle = false;
+
+} // namespace
+
+void ToggleItemLabelHighlight()
+{
+	invertHighlightToggle = !invertHighlightToggle;
+}
+
+void AltPressed(bool pressed)
+{
+	altPressed = pressed;
+}
+
+bool IsGeneratingItemLabels()
+{
+	return generatingLabelsState == 1;
+}
+
+bool IsItemLabelHighlighted()
+{
+	return isLabelHighlighted;
+}
+
+void UpdateItemLabelOffsets(const CelOutputBuffer &out, BYTE *dst, int width)
+{
+	int xval = (dst - out.begin()) % out.pitch();
+	if (xval < drawMinX)
+		drawMinX = xval;
+	xval += width;
+	if (xval > drawMaxX)
+		drawMaxX = xval;
+}
+
+void GenerateItemLabelOffsets(const CelOutputBuffer &out)
+{
+	if (generatingLabelsState == 2)
+		return;
+	generatingLabelsState = 1;
+	int itemTypes = gbIsHellfire ? ITEMTYPES : 35;
+	for (int i = 0; i < itemTypes; i++) {
+		drawMinX = gnScreenWidth;
+		drawMaxX = 0;
+		CelClippedDrawTo(out, out.pitch() / 2 - 16, 351, itemanims[i], ItemAnimLs[i], 96);
+		labelCenterOffsets[i] = drawMinX - out.pitch() / 2 + (drawMaxX - drawMinX) / 2;
+	}
+	generatingLabelsState = 2;
+}
+
+void AddItemToLabelQueue(int x, int y, int id)
+{
+	if (altPressed == invertHighlightToggle)
+		return;
+	ItemStruct *it = &items[id];
+
+	char textOnGround[64];
+	if (it->_itype == ITYPE_GOLD)
+		sprintf(textOnGround, "%i gold", it->_ivalue);
+	else
+		sprintf(textOnGround, "%s", it->_iIdentified ? it->_iIName : it->_iName);
+
+	int nameWidth = GetTextWidth((char *)textOnGround);
+	x += labelCenterOffsets[ItemCAnimTbl[it->_iCurs]];
+	y -= TILE_HEIGHT;
+	if (!zoomflag) {
+		x <<= 1;
+		y <<= 1;
+	}
+	x -= nameWidth >> 1;
+	text_color clr = COL_WHITE;
```

Comment:
Make a function along these lines:

```c++
text_color GetTextColor(const ItemStruct &it) {
	switch (it._iMagical) {
		case ITEM_QUALITY_MAGIC:
			return COL_BLUE;
		case ITEM_QUALITY_UNIQUE:
			return COL_GOLD
		default:
			return COL_WHITE;
	}
}
```

Then here:

```suggestion
	const text_color clr = GetTextColor(*it);
```

Reply:
I think I actually did recently! :D